src/indexing.jl

```diff
@@ -18,8 +18,10 @@ struct IndexEntry
 end
 
 IndexEntry(file) = IndexEntry(Path(file))
-function IndexEntry(filepath::AbstractPath)
-    isfile(filepath) || throw(DomainError(filepath, "Need an existant file."))
+function IndexEntry(filepath::AbstractPath, check_exists=false)
```

Comment:
Isn't this an internal method?

if that is the case then I am not sure that it sense to give it an option that is always set to false internally.

We can just never check `isfile`

Reply:
`IndexEntry` is [exported](https://github.com/invenia/Checkpoints.jl/blob/708cde9c3e6bd4e3b25bad15a992363a54c4ae82/src/Checkpoints.jl#L23), but it's true that `check_exists` can't be controlled at the level of `index_checkpoint_files` with the current changes.

If `check_exists=false` then this is also a breaking change.

Reply:
If noone relies on it. then it isn't really breaking.

Reply:
why would we ever want to check for individual files in `index_checkpoint_files`?

Reply:
Happy to change default constructor to check by default -- yay or nay?

Reply:
> If noone relies on it. then it isn't really breaking.
> We can be practical about it.

I'm surprised reading that - the last couple of times breaking-ness has been a grey area, I've erred on the side of breaking because it's hard to anticipate all the effects. But still inclined to agree - indexing is relatively new and I don't expect anyone to be relying on the error-checking when constructing `IndexEntry` directly.

On the other hand `index_checkpoint_files` shouldn't break. Is it the case that, if `isdir(dir)` in `index_checkpoint_files(dir)`, then `isfile(filepath)` for all `filepath` in `walkpath(dir)`? If so, we can remove the check of `isfile` here.

Reply:
> Is it the case that, if isdir(dir) in index_checkpoint_files(dir), then isfile(filepath) for all filepath in walkpath(dir)

Need to add that `filepath` is filtered to JLSO files. It's definitely true for `index_files` because that filters directly by `isfile`.

Reply:
> Happy to change default constructor to check by default -- yay or nay?

I still say we can remove it outright.
It is basically just a check to make sure the end-user is not doing something silly.


>>    If noone relies on it. then it isn't really breaking.
>>    We can be practical about it.
>
> I'm surprised reading that - the last couple of times breaking-ness has been a grey area, I've erred on the side of breaking because it's hard to anticipate all the effects. But still inclined to agree - indexing is relatively new and I don't expect anyone to be relying on the error-checking when constructing IndexEntry directly.

Hitting the right level of pragmatism on breaking or not breaking is nontrivial, and need context.
Like you say, having the context that this is relatively new and noone is probably using it directly.
It's not really even useful to use directly.
Probably, this shouldn't even have been exported. That was probably a mistake on my part.

ColPrac has in it's [list of changes that are not considered breaking](https://github.com/SciML/ColPrac#changes-that-are-not-considered-breaking):

> Exception behavior:
>
>    Exceptions may be replaced with non-error behavior. For instance, we may change a function to compute a result instead of raising an exception, even if that error is documented.

What this change is effectly doing is allowing the thing that was index to be something other than a file.
It might be a directory.
We don't ever want that right now.

---

> Need to add that filepath is filtered to JLSO files. 

In particular it is filtered to thing with the `jlso` extension.
Technically, a folder can have that extension.
And `walkpath` on normal filesystems does return leaf folders also.
but we know that never happens in the wild as we would have got complaints.
It never happens on S3 filesystems, as S3 has not representation of a empty folder, it can't be constructed.
(Because S3 doesn't really have folders, it has prefixes in a key)

---

Anyway, all in all I think deleting this check and not giving an option for it is worth it, and should not be considered a breaking change.
I am willing to give up a lot of purity of correctness in exchange for 75x speedup.
And adding a feature that is not going to be used right now is premature. We can easily add it later.
And later we will have more information on exactly what is needed.

Reply:
> We could check for / endings and filter them out to make index_files fast, too.

I think leave it alone.
It has it's own separate use-case in Power Simulations land.
Which doesn't involved S3 right now AFAIK ( cc @nickrobinson251 )

More generally though, maybe FilePathsBase needs a sub-API for pure operations that work without touching the file-system, like python's pathlib has. (cc @rofinn)
so `isfile_pure`  could on `S3Path`'s always return `true` since we know it is impossible to represent anything else, if you don't also need to check it it exists.

Reply:
> ColPrac has in it's list of changes that are not considered breaking:

Thanks for explaining, I was not aware of that guideline. 

> Anyway, all in all I think deleting this check and not giving an option for it is worth it, and should not be considered a breaking change.

I agree with this now.

Reply:
What about specialising `index_files` for S3Paths (cf latest commit)

Reply:
I forgot to mention that crucially `index_checkpoint_files` has already checked if files are .jlso's so in particular they are files. The purpose of this MR is purely to remove the redundant check, but now that I've noticed the slow checks on `index_files`, it makes sense to fix that as well. 

If this is a separate MR let me know..

Reply:
>so isfile_pure could on S3Path's always return true since we know it is impossible to represent anything else, if you don't also need to check it it exists.

`walkpath` on S3 has been changed(?) to include directories, I believe for consistency with local paths.

Reply:
> If this is a separate MR let me know..

This *is* a separate MR.

Reply:
>But I can name a folder foo.jlso 
👎 
Reverted to simply skipping checks for `IndexEntry`, and throwing ArgumentError if `dir` in `index_checkpoint_files(dir)` is missing